### ISSUE TITLE
Better type handling?

### ISSUE THREAD

**Issue by jakewilliami** [JuliaImages/IntegralArrays.jl#15]
If you have an array of Int8, it is somewhat likely that the bottom right corner of the array will not fit into 8 bits.  Should we handle addition overflow better?
```julia
julia> iX = IntegralArray(rand(Int8, 10, 10))
ERROR: InexactError: trunc(Int8, 145)
```

**Comment by timholy** [comment]
Good idea. Proposal: allow the user to specify the type, `IntegralArray{T}(A)`. Then https://github.com/JuliaImages/IntegralArrays.jl/blob/9ded0c8839f4205be9577c4da163feaf282da6d0/src/IntegralArrays.jl#L97 doesn't call `similar` itself (that happens in `IntegralArray{T}(A)`), it just fills in the type and calls `IntegralArray{T}(A)`. Finally, we could check for small integers (see https://github.com/JuliaLang/julia/blob/e0a4b7727c2760a61c69c69c1737c46e32223d4f/base/range.jl#L799-L801). Care to give that a try?

**Comment by jakewilliami** [comment]
Absolutely, I'll give it ago!

P.S., this is such a well-made package.  I was prepared to have a go at implementing [n-dimensional support for this](https://www.sciencedirect.com/science/article/abs/pii/S0167865510003533) but it's already been done!  And it supports summing regions via IntervalSets.jl.  Great work on this.

**Comment by jakewilliami** [comment]
I made a commit (a703897).  Not sure if we want to also have safe addition on `IntegralArray(data::AbstractArray, A::AbstractArray)`, or if it's up to the user to ensure `data` has the corret type?  In my commit I've attempted to fix both methods, but I'm not sure if converting `data` as I have implemented is optimal.

**Comment by timholy** [comment]
You may have seen I left a couple of comments. With the dual-array constructor, we use whatever `data` the user provides.That's more of an internal constructor, of course. Perhaps it should actually be named `IntegralArray!(data, A)`?

**Comment by jakewilliami** [comment]
Thank you for your comments!  I've implemented your comments; see 432e574—is this what you mean?  Once again, I was unsure how exactly one would change the `smallints` stuff, so please let me know what you think.

I was wondering about the bang convention too, and I think that would be a good change to make (stick to convention), but would like to get the final say from @johnnychen94, as I notice he's done a lot of work on this package.  The question I have around implementation of this: the reason we can use parametric constructors (`IntegralArray{T}`) is because `IntegralArray` is a `struct`.  So should we be adding an argument for `IntegralArray!` (as we cannot type `IntegralArray!{T}`); e.g.:
```julia
function IntegralArray!(T::Type, data::AbstractArray, A::AbstractArray)
    # ...
end
```

**Comment by johnnychen94** [comment]
I don't think we have the bang conversions for types, `IntegralArray{T}` is how user-specified eltype is typically supported for custom array objects. @jakewilliami would you mind opening a PR for this as that would be easier to review and discuss.
